File: lintcode/python/0617_maximum_average_subarray_II.py
#!/usr/bin/python -t

# binary search
#使用九章算法强化班中讲过的基于二分答案的方法
#二分出 average 之后，把数组中的每个数都减去 average，然后的任务就是去求这个数组中，是否有长度 >= k 的 subarray，他的和超过 0。
#这一步用类似 Maximum Subarray 的解法来做就好了

class Solution:
    # @param {int[]} nums an array with positive and negative numbers
    # @param {int} k an integer
    # @return {double} the maximum average
    def maxAverage(self, nums, k):
        csum = [0]
        for i in range(len(nums)):
            csum.append(csum[i]+nums[i])

        print csum
        res = float(csum[k])/k
        check = 0
        for i in range(k+1,len(csum)):
            left = float((csum[i] - csum[check]))/(i - check)
            right = float((csum[i] - csum[i-k]))/k
            print i, check, left, right
            if left >= right:
                if left > res:
                    res = left
            else:
                if right > res:
                    res = right
                check = i - k
        return res

if __name__ == '__main__':
    s = [-4,5,-4,5,-4,5,-4,5,-4,5,-4,5,-4,5,-4,5,-4,5,-1000]
    t = 10
    ss = Solution()
    print "answer is\n"
    print ss.maxAverage(s, t)

File: lintcode/golang/0617_maximum_average_subarray_II.go

// should be correct, but not AC
/**
 * @param nums: an array with positive and negative numbers
 * @param k: an integer
 * @return: the maximum average
 */
 
import "fmt"

func maxAverage (nums []int, k int) float64 {
    // write your code here
    n := len(nums)
    if n == 0 {
        return 0
    }
    
    l := min(nums)
    r := max(nums)
    
    for l + 0.0001 < r {
        mid := (l+r)/2
        
        if canFind(nums, k, mid) {
            l = mid
        } else {
            r = mid
        }
    }
    
    return l
}

func max(nums []int) float64 {
    var ret int
    ret = nums[0]
    
    for _, n := range nums {
        if n > ret {
            ret = n
        }
    }
    return float64(ret)
}

func min(nums []int) float64 {
    var ret int
    ret = nums[0]
    
    for _, n := range nums {
        if n < ret {
            ret = n
        }
    }
    return float64(ret)
}

func canFind(nums []int, k int, target float64) bool {
    pre_sum := []float64{0}
    
    for _, v := range nums {
        pre_sum = append(pre_sum, pre_sum[len(pre_sum)-1] + float64(v)- target)
    }
    
    var min_pre float64
    
    for i:= k; i < len(nums)+ 1; i++ {
        if pre_sum[i] - min_pre >= 0 {
            return true
        }
        fmt.Println(i, k, pre_sum)
        min_pre = min_float64(min_pre, pre_sum[i-k+1])
    }
    
    return false
}

func min_float64(a, b float64) float64 {
    if a < b {
        return a
    }
    return b
}
